sqlhooks.go

```diff
@@ -59,6 +66,68 @@ func (conn *Conn) Prepare(query string) (driver.Stmt, error) { return conn.Conn.
 func (conn *Conn) Close() error                              { return conn.Conn.Close() }
 func (conn *Conn) Begin() (driver.Tx, error)                 { return conn.Conn.Begin() }
 
+// ExecerContext implements a database/sql.driver.ExecerContext
+type ExecerContext struct {
+	*Conn
+}
+
+func isExecer(conn driver.Conn) bool {
+	switch conn.(type) {
+	case driver.ExecerContext:
+		return true
+	case driver.Execer:
+		return true
+	default:
+		return false
+	}
+}
+
+func (conn *ExecerContext) execContext(ctx context.Context, query string, args []driver.NamedValue) (driver.Result, error) {
+	switch c := conn.Conn.Conn.(type) {
+	case driver.ExecerContext:
+		return c.ExecContext(ctx, query, args)
+	case driver.Execer:
+		dargs, err := namedValueToValue(args)
+		if err != nil {
+			return nil, err
+		}
+		return c.Exec(query, dargs)
+	default:
+		// This should not happen
+		return nil, errors.New("ExecerContext created for a non Execer driver.Conn")
+	}
+}
+
+func (conn *ExecerContext) ExecContext(ctx context.Context, query string, args []driver.NamedValue) (driver.Result, error) {
+	var err error
+
+	list := namedToInterface(args)
+
+	// Exec `Before` Hooks
+	if ctx, err = conn.hooks.Before(ctx, query, list...); err != nil {
+		return nil, err
+	}
+
+	results, err := conn.execContext(ctx, query, args)
+	if err != nil {
+		return results, err
+	}
+
+	if ctx, err = conn.hooks.After(ctx, query, list...); err != nil {
+		return nil, err
+	}
+
+	return results, err
+}
+
+func (conn *ExecerContext) Exec(query string, args []driver.Value) (driver.Result, error) {
+	// We have to implement Exec since it is required in the current version of
+	// Go for it to run ExecContext. From Go 10 it will be optional. However,
+	// this code should never run since database/sql always prefers to run
+	// ExecContext.
+	return nil, errors.New("Exec was called when ExecContext was implemented")
+}
```

Comment:
@gchaincl on second thought, if you are running sqlhooks pre go1.8 this will be called which is bad.